File: Python/BST-1.py
class Node:
    def __init__(self, key: int, name: str):
        self.key = key
        self.name = name

        self.left = None
        self.right= None

    def __str__(self):
        return "%s has a key %d" %(self.name, self.key)


class BinaryTree:
    def __init__(self):
        self.root = None

    def addNode(self, key: int, name: str) -> None:
        newNode = Node(key, name)

        if self.root is None:
            self.root = newNode
        else:
            focus = self.root
            parent = None

            while True:
                parent = focus

                if key < focus.key:
                    focus = focus.left

                    if focus is None:
                        parent.left = newNode
                        return
                else:
                    focus = focus.right

                    if focus is None:
                        parent.right = newNode
                        return

    def inOrder(self, focus: Node):
        if focus is not None:
            self.inOrder(focus.left)
            print(focus)
            self.inOrder(focus.right)

    def preOrder(self, focus: Node):
        if focus is not None:
            print(focus)
            self.preOrder(focus.left)
            self.preOrder(focus.right)

    def postOrder(self, focus: Node):
        if focus is not None:
            self.postOrder(focus.left)
            self.postOrder(focus.right)
            print(focus)

    def findNode(self, key: Node) -> Node:
        focus = self.root

        while focus.key != key :
            if key < focus.key :
                focus = focus.left
            else:
                focus = focus.right

            if focus is None:
                return None
        return focus


if __name__ == '__main__':
    tree = BinaryTree()

    tree.addNode(50, "Boss")
    tree.addNode(25, "Vice President")
    tree.addNode(15, "Office Manager")
    tree.addNode(30, "Secretary")
    tree.addNode(75, "Sales Manager")
    tree.addNode(85, "Salesman 1")

    # tree.inOrder(tree.root)
    # tree.preOrder(tree.root)
    tree.postOrder(tree.root)

    print("", tree.findNode(30))

File: Ruby/BST-1.rb
class BinaryTree
    attr_reader :root
    root = nil

    def addNode(key, name)
        newNode = Node.new(key, name)

        if(root == nil)
            @root = newNode
        else
            focus = @root
            parent = nil

            while true
                parent = focus

                if key < focus.key
                    focus = focus.leftChild

                    if focus == nil
                        parent.leftChild = newNode
                        return
                    end
                else
                    focus = focus.rightChild

                    if focus == nil
                        parent.rightChild = newNode
                        return
                    end
                end
            end
        end
    end

    def inOrder(focus)
        if focus != nil
            inOrder(focus.leftChild)
            puts(focus)
            inOrder(focus.rightChild)
        end
    end

    def preOrder(focus)
        if focus != nil
            puts(focus)
            preOrder(focus.leftChild)
            preOrder(focus.rightChild)
        end
    end

    def postOrder(focus)
        if focus != nil
            postOrder(focus.leftChild)
            postOrder(focus.rightChild)
            puts(focus)
        end
    end

    def find(key)
        focus = @root

        while focus.key != key
            if key < focus.key
                focus = focus.leftChild
            else
                focus = focus.rightChild
            end

            if focus == nil
                return nil
            end
        end
        return focus
    end
end


class Node
    attr_accessor :key, :name, :leftChild, :rightChild
    
    def initialize(key, name)
        @key = key
        @name = name
    end

    def to_s()
        return "#{@name} has a key #{@key}"
    end
end


if __FILE__ == $0
    tree = BinaryTree.new()

    tree.addNode(50, "Boss")
    tree.addNode(25, "Vice President")
    tree.addNode(15, "Office Manager")
    tree.addNode(30, "Secretary")
    tree.addNode(75, "Sales Manager")
    tree.addNode(85, "Salesman 1")

    # tree.inOrder(tree.root)
    # tree.preOrder(tree.root)
    tree.postOrder(tree.root)

    puts(tree.find(30))
end
